Sources/MessagingPush/PushClickHandler.swift

```diff
@@ -0,0 +1,213 @@
+        guard let parsedPush = CustomerIOParsedPushPayload.parse(response: response, jsonAdapter: jsonAdapter) else {
+            // push not sent from CIO. exit early
+            return nil
+        }
+
+        if response.didClickOnPush {
+            pushClicked(response, parsedPush: parsedPush)
+        }
+
+        return parsedPush
+    }
+
+    func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) -> Bool {
+        guard let parsedPush = CustomerIOParsedPushPayload.parse(response: response, jsonAdapter: jsonAdapter) else {
+            // push did not come from CIO
+            // Do not call completionHandler() because push did not come from CIO. Another service might have sent it so
+            // allow another SDK
+            // to call the completionHandler()
+            return false
+        }
+
+        if response.didClickOnPush {
+            pushClicked(response, parsedPush: parsedPush)
+        }
+
+        completionHandler()
+        return true
+    }
+
+    private func pushClicked(_ response: UNNotificationResponse, parsedPush: CustomerIOParsedPushPayload) {
+        // TODO: prevent duplicate push metrics and deep link handling.
+
+        // Now we are ready to handle the push click.
+        // Track metrics
+        if sdkConfig.autoTrackPushEvents {
+            customerIO?.trackMetric(deliveryID: parsedPush.deliveryId, event: .opened, deviceToken: parsedPush.deviceToken)
+        }
+
+        // Cleanup files on device that were used when the push was displayed. Files are no longer
+        // needed now that the push is no longer shown.
+        cleanupAfterPushInteractedWith(pushContent: parsedPush)
+
+        // Handle deep link, if there is one attached to push.
+        if let deepLinkUrl = parsedPush.deepLink {
+            deepLinkUtil.handleDeepLink(deepLinkUrl)
+        }
+    }
+
+    // There are files that are created just for displaying a rich push. After a push is interacted with, those files
+    // are no longer needed.
+    // This function's job is to cleanup after a push is no longer being displayed.
+    func cleanupAfterPushInteractedWith(pushContent: CustomerIOParsedPushPayload) {
+        pushContent.cioAttachments.forEach { attachment in
+            let localFilePath = attachment.url
+
+            try? FileManager.default.removeItem(at: localFilePath)
+        }
+    }
+}
+
+// Swizzle functions
+extension PushClickHandlerImpl {
+    // Swizzled method that gets called when a new UNUserNotificationCenter.delegate gets set.
+    @objc dynamic func cio_swizzled_setDelegate(delegate: UNUserNotificationCenterDelegate?) {
+        guard let delegate = delegate else {
+            cio_swizzled_setDelegate(delegate: delegate) // continue swizzle
```

Comment:
Explanation of `continue swizzling` is in another branch [here](https://github.com/customerio/customerio-ios/commit/6050e7b6cb4f3329c7a1b539ed14e0e7970eae3a#diff-aafef34d456fef5f84dba8d1dc5f3751b89d3dcdedfbfa1572d4806b8f70b426R3-R22).